### PR TITLE
refactor!: added ability to serialize/deserialize pods list and content items

### PR DESCRIPTION
**Pull request by IgorShadurin**
This PR changes the return data format for some methods. In general, all methods now return plain objects/arrays for ease of conversion to/from JSON.

Changed methods:
`fdp.directory.read`
`fdp.file.downloadData`
`fdp.personalStorage.list`
`fdp.personalStorage.create`

Close https://github.com/fairDataSociety/fdp-storage/issues/209

### CODE REVIEW

**Review by nugaon — COMMENTED** [review]
See my comments below, but it can be merged from my side after @tomicvladan approved.

File: src/content-items/serialization.ts
Comment (IgorShadurin) [reply src/content-items/serialization.ts:18]: Created an issue for this improvement for a next iteration https://github.com/fairDataSociety/fdp-storage/issues/215

**Comment by tomicvladan** [comment]
This approach could work, but it will be still complicated to use. Because there is no uniform API. The proxy object in Blossom will have to intercept each call and analyze whether it should apply serialization/deserialization or not. This might be difficult to maintain later when new changes are applied to fdp-storage. It will be difficult to keep track of changes and whether they require updates on the proxy side.

Instead of that, I think much better approach would be to reorganize these two classes: `DirectoryItem` and `List`. In the `DirectoryItem` class, instead of having `getFiles` and `getDirectories` methods, the class could contain public properties `files` and `directories` already filtered. It won't increase memory too much because those are just two arrays with the same references. But there will be no need for any methods inside it. Similar can be done with the `List` as well.

In the future we can keep the same approach for new features, so each method returns only pure objects without any methods inside. I think that will make fdp-storage much more usable, especially because we plan to use it in a variety of environments.

**Comment by IgorShadurin** [comment]
> This approach could work, but it will be still complicated to use. Because there is no uniform API. The proxy object in Blossom will have to intercept each call and analyze whether it should apply serialization/deserialization or not. This might be difficult to maintain later when new changes are applied to fdp-storage. It will be difficult to keep track of changes and whether they require updates on the proxy side.
> 
> Instead of that, I think much better approach would be to reorganize these two classes: `DirectoryItem` and `List`. In the `DirectoryItem` class, instead of having `getFiles` and `getDirectories` methods, the class could contain public properties `files` and `directories` already filtered. It won't increase memory too much because those are just two arrays with the same references. But there will be no need for any methods inside it. Similar can be done with the `List` as well.
> 
> In the future we can keep the same approach for new features, so each method returns only pure objects without any methods inside. I think that will make fdp-storage much more usable, especially because we plan to use it in a variety of environments.

@nugaon wdyt?

**Comment by nugaon** [comment]
I agree with @tomicvladan let's rid of the classes and make simple objects/arrays from those.

**Comment by tomicvladan** [comment]
I think all the changes should be:

### `src/pod/list.ts`

Instead of:

```typescript
getPods(): Pod[];
getSharedPods(): SharedPod[];
```

Should be:


```typescript
pods: Pod[];
sharedPods : SharedPod[];
```

### `src/content-items/directory-item.ts`

Instead of:

```typescript
getFiles(): FileItem[];
getDirectories(): DirectoryItem[];
static fromRawDirectoryMetadata(item: RawDirectoryMetadata): DirectoryItem;
```

Should be:


```typescript
files: FileItem[];
directories: DirectoryItem[];

// maybe exported function form another module?
export function fromRawDirectoryMetadata(item: RawDirectoryMetadata): DirectoryItem;
```

**Comment by tomicvladan** [comment]
I remembered one more class that needs to be handled somehow, the `Data` response returned by `File.downloadData`. It has methods that transform data to different forms.

**Comment by IgorShadurin** [comment]
Now the returned results are simple objects that can be converted via JSON.stringify/parse

@tomicvladan

**Review by nugaon — COMMENTED** [review]
We need to have a breaking change and make all returns JSON compatible without any mapping and transitions.
@tomicvladan, approve this PR if everything goes smoothly in Blossom.

File: src/pod/utils.ts
Comment (tomicvladan) [reply src/pod/utils.ts:184]: Yes, the idea is to make a breaking change to the API, so when using the library in proxy, users won't notice that they are using proxy. To achieve that we shouldn't have any kind of serialization/deserialization. Just all objects that are returned from methods need to be simple objects without any methods inside them.

File: src/content-items/serialization.ts
Comment (tomicvladan) [inline src/content-items/serialization.ts:6]: This file can be deleted as well?

Comment (tomicvladan) [reply src/content-items/serialization.ts:6]: Sorry, I was confused by the classes names. Do we need the `Serializable` suffix? It might confuse regular users.

**Comment by tomicvladan** [comment]
Other than class names, it looks good. I just think that we need to remove the `Serializable` from all class names.

**Comment by nugaon** [comment]
before merge, please update PR details about the interface changes

**Review by tomicvladan — APPROVED** [review]

**Pull request merged**
